File: packages/allstak-python/e2e_test.py
"""
End-to-end test for AllStak Python SDK against localhost:8080.

Tests: tracing (spans), logs, HTTP requests, error capture.
"""

import time
import allstak

# ---------------------------------------------------------------------------
# 1. Initialize the SDK
# ---------------------------------------------------------------------------

SERVICE = "sdk-python-test"
ENVIRONMENT = "e2e-testing"

client = allstak.init(
    api_key="ask_live_4574x2yao33rtjbiuf2q873ltv6vpokb",
    host="http://localhost:8080",
    environment=ENVIRONMENT,
    debug=True,
    max_retries=1,
    flush_interval_ms=500,
    connect_timeout=5.0,
    read_timeout=5.0,
)

# Set service name on the tracing module
client.tracing.set_service(SERVICE)

print(f"[OK] SDK initialized (host=http://localhost:8080, service={SERVICE})")

# ---------------------------------------------------------------------------
# 2. Distributed Tracing - root + child span
# ---------------------------------------------------------------------------

# Start root span
root_span = allstak.start_span(
    "HTTP GET /api/products",
    description="Fetch product catalog",
    tags={"http.method": "GET", "http.url": "/api/products"},
)
trace_id = root_span.trace_id
print(f"[OK] Root span started: trace_id={trace_id}, span_id={root_span.span_id}")

# Simulate some work
time.sleep(0.05)

# Start child span
child_span = allstak.start_span(
    "DB SELECT products",
    description="SELECT * FROM products WHERE active=true",
    tags={"db.type": "postgresql", "db.statement": "SELECT"},
)
print(f"[OK] Child span started: span_id={child_span.span_id}, parent={child_span.parent_span_id}")

# Simulate DB query
time.sleep(0.03)

# Finish child span
child_span.finish("ok")
print("[OK] Child span finished")

# Finish root span
time.sleep(0.02)
root_span.finish("ok")
print("[OK] Root span finished")

# ---------------------------------------------------------------------------
# 3. Logs with trace_id
# ---------------------------------------------------------------------------

allstak.log.info(
    "Product catalog fetched successfully",
    service=SERVICE,
    trace_id=trace_id,
    environment=ENVIRONMENT,
    metadata={"product_count": 42},
)

allstak.log.warn(
    "Slow query detected in products table",
    service=SERVICE,
    trace_id=trace_id,
    environment=ENVIRONMENT,
    metadata={"query_ms": 3200, "table": "products"},
)

allstak.log.error(
    "Cache miss for product catalog",
    service=SERVICE,
    trace_id=trace_id,
    environment=ENVIRONMENT,
    metadata={"cache_key": "products:active:v2"},
)

print(f"[OK] 3 log messages queued (trace_id={trace_id})")

# ---------------------------------------------------------------------------
# 4. HTTP Request Records (3 inbound requests, varying methods/statuses)
# ---------------------------------------------------------------------------

allstak.http.record(
    direction="inbound",
    method="GET",
    host="api.myapp.com",
    path="/api/products",
    status_code=200,
    duration_ms=85,
    request_size=0,
    response_size=4096,
    trace_id=trace_id,
)

allstak.http.record(
    direction="inbound",
    method="POST",
    host="api.myapp.com",
    path="/api/orders",
    status_code=201,
    duration_ms=120,
    request_size=512,
    response_size=256,
    trace_id=trace_id,
)

allstak.http.record(
    direction="inbound",
    method="DELETE",
    host="api.myapp.com",
    path="/api/products/99",
    status_code=404,
    duration_ms=12,
    request_size=0,
    response_size=64,
    trace_id=trace_id,
)

allstak.http.record(
    direction="outbound",
    method="GET",
    host="cache.redis.internal",
    path="/products/catalog",
    status_code=200,
    duration_ms=3,
    trace_id=trace_id,
)

print("[OK] 4 HTTP request records queued")

# ---------------------------------------------------------------------------
# 5. Error / Exception capture with trace context
# ---------------------------------------------------------------------------

try:
    # Simulate a real exception
    data = {"products": None}
    count = len(data["products"]["items"])  # TypeError
except Exception as e:
    event_id = allstak.capture_exception(
        e,
        level="error",
        environment=ENVIRONMENT,
        metadata={
            "source": "e2e_test",
            "operation": "product_count",
            "traceId": trace_id,
        },
    )
    print(f"[OK] Exception captured -> event_id={event_id}")

# ---------------------------------------------------------------------------
# 6. Flush all data
# ---------------------------------------------------------------------------

print("\nFlushing all buffers...")
allstak.flush()
time.sleep(1)  # give background workers time to complete
allstak.flush()  # second flush to catch stragglers
time.sleep(0.5)

print("\n" + "=" * 60)
print("E2E TEST COMPLETE")
print("=" * 60)
print(f"  Trace ID:    {trace_id}")
print(f"  Service:     {SERVICE}")
print(f"  Environment: {ENVIRONMENT}")
print(f"  Spans sent:  2 (root + child)")
print(f"  Logs sent:   3 (info, warn, error)")
print(f"  HTTP reqs:   4 (3 inbound + 1 outbound)")
print(f"  Errors:      1 (TypeError)")
print("=" * 60)

File: packages/allstak-js/e2e_test.mjs
/**
 * AllStak JS SDK — E2E Test
 *
 * Tests spans, logs, HTTP requests, and error capture against localhost:8080.
 */

import { AllStak } from './dist/node/index.mjs';

const API_KEY = 'ask_live_4574x2yao33rtjbiuf2q873ltv6vpokb';
const BASE_URL = 'http://localhost:8080';

// DSN format: http://<apiKey>@<host>
const DSN = `http://${API_KEY}@localhost:8080`;

console.log('=== AllStak JS SDK E2E Test ===\n');
console.log(`DSN: ${DSN}`);

// --- Initialize ---
const client = AllStak.init({
  dsn: DSN,
  environment: 'e2e-testing',
  release: '0.1.0-e2e',
  tags: { service: 'sdk-js-test' },
});

AllStak.setUser({ id: 'e2e-user-1', email: 'e2e@test.com' });

console.log(`Session ID: ${AllStak.getSessionId()}`);

// --- Step 1: Create Spans ---
console.log('\n--- Step 1: Tracing (Spans) ---');

const rootSpan = AllStak.startSpan('HTTP GET /api/dashboard', {
  description: 'Handle dashboard API request',
  tags: { 'http.method': 'GET', 'http.route': '/api/dashboard' },
});

console.log(`Trace ID : ${rootSpan.traceId}`);
console.log(`Root Span: ${rootSpan.spanId} (operation: HTTP GET /api/dashboard)`);

// Simulate some work
await sleep(50);

// Child span
const childSpan = AllStak.startSpan('API fetch_metrics', {
  description: 'Fetch metrics from metrics service',
  tags: { 'service.target': 'metrics-service' },
});
console.log(`Child Span: ${childSpan.spanId} (operation: API fetch_metrics)`);

await sleep(30);

// Finish child first, then root
childSpan.finish('ok');
console.log('  -> child span finished');

await sleep(20);

rootSpan.finish('ok');
console.log('  -> root span finished');

// --- Step 2: Logs with trace context ---
console.log('\n--- Step 2: Logs ---');

// Start a new span so logs get trace context
const logSpan = AllStak.startSpan('log-context-span', {
  description: 'Span for log context',
});

AllStak.log.info('Dashboard loaded successfully', {
  service: 'sdk-js-test',
  userId: 'e2e-user-1',
  dashboard: 'main',
});
console.log('  -> sent info log');

AllStak.log.warn('Slow query detected in metrics fetch', {
  service: 'sdk-js-test',
  queryTime: 450,
});
console.log('  -> sent warn log');

AllStak.log.error('Failed to load widget data', {
  service: 'sdk-js-test',
  widget: 'revenue-chart',
  error: 'TIMEOUT',
});
console.log('  -> sent error log');

logSpan.finish('ok');

// --- Step 3: HTTP Request Records ---
console.log('\n--- Step 3: HTTP Requests ---');

AllStak.captureRequest({
  direction: 'inbound',
  method: 'GET',
  host: 'api.myapp.com',
  path: '/api/dashboard',
  statusCode: 200,
  durationMs: 142,
  requestSize: 0,
  responseSize: 8192,
  userId: 'e2e-user-1',
});
console.log('  -> captured GET /api/dashboard (200, 142ms)');

AllStak.captureRequest({
  direction: 'outbound',
  method: 'POST',
  host: 'metrics.internal',
  path: '/v1/query',
  statusCode: 200,
  durationMs: 89,
  requestSize: 256,
  responseSize: 4096,
});
console.log('  -> captured POST /v1/query (200, 89ms)');

AllStak.captureRequest({
  direction: 'inbound',
  method: 'GET',
  host: 'api.myapp.com',
  path: '/api/users/42',
  statusCode: 404,
  durationMs: 12,
  requestSize: 0,
  responseSize: 64,
  userId: 'e2e-user-1',
});
console.log('  -> captured GET /api/users/42 (404, 12ms)');

// --- Step 4: Capture an Error ---
console.log('\n--- Step 4: Error Capture ---');

const errSpan = AllStak.startSpan('error-context-span');
try {
  throw new Error('E2E test: NullPointerException in UserService.getProfile');
} catch (err) {
  AllStak.captureException(err, {
    service: 'sdk-js-test',
    module: 'user-service',
    action: 'getProfile',
  });
  console.log(`  -> captured error: ${err.message}`);
}
errSpan.finish('error');

// --- Step 5: Flush & Destroy ---
console.log('\n--- Step 5: Flushing ---');

// The SDK uses internal timers. Destroy triggers a final flush.
// Give a moment for any in-flight sends, then destroy.
await sleep(1000);
AllStak.destroy();
console.log('  -> SDK destroyed (final flush triggered)');

// Wait for network requests to complete
await sleep(2000);

// --- Summary ---
const traceId = rootSpan.traceId;
console.log('\n=== E2E Test Complete ===');
console.log(`Trace ID: ${traceId}`);
console.log('Sent:');
console.log('  - 3 spans (root + child + log-context)');
console.log('  - 3 logs (info, warn, error) with trace context');
console.log('  - 3 HTTP request records');
console.log('  - 1 captured error with trace context');

// --- Step 6: Verify in ClickHouse ---
console.log('\n--- Step 6: ClickHouse Verification ---');

try {
  const chQuery = `SELECT trace_id, operation, duration_ms, service, environment FROM allstak.spans WHERE service='sdk-js-test' ORDER BY start_time DESC LIMIT 10 FORMAT PrettyCompact`;
  const chRes = await fetch(`http://localhost:8123/?query=${encodeURIComponent(chQuery)}`);
  const chText = await chRes.text();
  if (chRes.ok && chText.trim().length > 0) {
    console.log('ClickHouse spans:\n' + chText);
  } else {
    console.log(`ClickHouse query status: ${chRes.status}`);
    console.log(chText || '(empty result — spans may not have been written yet)');
  }
} catch (err) {
  console.log(`ClickHouse not reachable: ${err.message}`);
}

try {
  const logQuery = `SELECT trace_id, level, message, service FROM allstak.logs WHERE service='sdk-js-test' ORDER BY timestamp DESC LIMIT 10 FORMAT PrettyCompact`;
  const logRes = await fetch(`http://localhost:8123/?query=${encodeURIComponent(logQuery)}`);
  const logText = await logRes.text();
  if (logRes.ok && logText.trim().length > 0) {
    console.log('ClickHouse logs:\n' + logText);
  } else {
    console.log(`ClickHouse logs query status: ${logRes.status}`);
    console.log(logText || '(empty result)');
  }
} catch (err) {
  console.log(`ClickHouse logs query failed: ${err.message}`);
}

console.log('\nDone.');

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
